File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
n = int(input())
res=[[0 for _ in range(10)] for _ in range(n+1)]
for i in range(1, 10):
    res[1][i]=1
for i in range(2, n+1):
    res[i][0]=res[i-1][1]
    res[i][9]=res[i-1][8]
    for j in range(1, 9):
        res[i][j]=res[i-1][j-1]+res[i-1][j+1]
sum = 0
for v in res[n]:
    sum+=v
print(sum%1000000000)

File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n int
	var sum uint64
	fmt.Fscanf(reader, "%d\n", &n)
	var res = make([][]uint64, n+1)
	for i := 0; i < n+1; i++ {
		res[i] = make([]uint64, 10)
	}
	for i := 1; i < 10; i++ {
		res[1][i] = 1
	}
	for i := 2; i < n+1; i++ {
		res[i][0] = res[i-1][1]
		res[i][9] = res[i-1][8]
		for j := 1; j < 9; j++ {
			res[i][j] = (res[i-1][j-1] + res[i-1][j+1]) % 1000000000
		}
	}
	for _, v := range res[n] {
		sum += v
	}
	fmt.Fprintf(writer, "%d\n", sum%1000000000)
}
